runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -969,6 +997,271 @@ J9::TransformUtil::canFoldStaticFinalField(TR::Compilation *comp, TR::Node* node
+   {
+   J9::TransformUtil::attemptStaticFinalFieldFoldingImpl(opt, currentTree, node, false);
+   }
+
+/** \brief
+ *     Try to fold static final field with protection
+ *
+ *  \param opt
+ *     The current optimization object.
+ *
+ *  \param currentTree
+ *     The tree with the load of static final field.
+ *
+ *  \param node
+ *     The node which is a load of a static final field.
+ *
+ *  \param varHandleOnly
+ *     True if only folding varHandle static final fields.
+ *     Faslse if folding all static final fileds.
+ */
+bool J9::TransformUtil::attemptStaticFinalFieldFoldingImpl(TR::Optimization* opt, TR::TreeTop * currentTree, TR::Node *node, bool varHandleOnly)
+   {
+   TR::Compilation* comp = opt->comp();
+   // first attempt folding reliable fields
+   if (J9::TransformUtil::foldReliableStaticFinalField(comp, node))
+      {
+      if (opt->trace())
+         traceMsg(comp, "SFFF fold reliable at node %p\n", node);
+      return true;
+      }
+
+   // try folding regular static final fields
+   if (comp->getOption(TR_DisableGuardedStaticFinalFieldFolding))
+      {
+      return false;
+      }
+
+   if (!comp->supportsInduceOSR()
+       || !comp->isOSRTransitionTarget(TR::postExecutionOSR)
+       || comp->getOSRMode() != TR::voluntaryOSR)
+      {
+      return false;
+      }
+
+   if (skipFinalFieldFoldingInBlock(comp, currentTree->getEnclosingBlock())
+       || safeToAddFearPointAt(opt, currentTree) != TR_yes
+       || J9::TransformUtil::canFoldStaticFinalField(comp, node) != TR_maybe)
+      {
+      return false;
+      }
+
+   TR::SymbolReference* symRef = node->getSymbolReference();
+   if (symRef->hasKnownObjectIndex())
+      {
+      return false;
+      }
+
+   int32_t cpIndex = symRef->getCPIndex();
+   TR_OpaqueClassBlock* declaringClass = symRef->getOwningMethod(comp)->getClassFromFieldOrStatic(comp, cpIndex);
+   int32_t fieldNameLen;
+   char* fieldName = symRef->getOwningMethod(comp)->fieldName(cpIndex, fieldNameLen, comp->trMemory(), stackAlloc);
+   int32_t fieldSigLength;
+   char* fieldSignature = symRef->getOwningMethod(comp)->staticSignatureChars(cpIndex, fieldSigLength);
+
+   if (opt->trace())
+      {
+      traceMsg(comp,
+              "Looking at static final field n%dn %.*s declared in class %p\n",
+              node->getGlobalIndex(), fieldNameLen, fieldName, declaringClass);
+      }
+   if (!varHandleOnly || isVarHandleFolding(comp, declaringClass, fieldSignature, fieldSigLength))
+      {
+      if (J9::TransformUtil::foldStaticFinalFieldAssumingProtection(comp, node))
+         {
+         // Add class to assumption table
+         comp->addClassForStaticFinalFieldModification(declaringClass);
+         // Insert osrFearPointHelper call
+         TR::TreeTop* helperTree = TR::TreeTop::create(comp,
+                                                    TR::Node::create(node,
```

Comment:
The indentation on these looks a bit weird can you check you are using spaces?

Reply:
Fixed indentation.